Correct event dispatching for multiple simultaneous touch points

mbrubeck opened the pull request:
Instead of just converting the mouse into a single "touch" input, Servo can now listen for multi-touch events from Glutin, maintain a list of active touch points, and dispatch events for all of them.

r? @glennw (for the compositor changes) and @jdm (for the DOM changes)

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/8232)

<!-- Reviewable:end -->

mbrubeck committed:
Handle multi-touch events from glutin

jdm commented:
+S-needs-code-changes 
<!-- Reviewable comment -K1ibK1sSk-QL6h-rz0_:1522908653 -->

---

Reviewed 3 of 7 files at r1, 1 of 1 files at r2.
Review status: 3 of 7 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

<sup>**[components/script/dom/document.rs, line 801 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8232#-K1ib6WMhEOuEve3jci8)** ([raw file](https://github.com/servo/servo/blob/be64b1c6b830444e51c84486ed571506f9782948/components/script/dom/document.rs#L801)):</sup>
`*t.Target() == target`
<!-- Reviewable comment -K1ib6WMhEOuEve3jci9:-49760716 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8232)
<!-- Sent from Reviewable.io -->

mbrubeck commented:
Review status: 2 of 7 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

<sup>**[components/script/dom/document.rs, line 801 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8232#-K1ib6WMhEOuEve3jci8)** ([raw file](https://github.com/servo/servo/blob/be64b1c6b830444e51c84486ed571506f9782948/components/script/dom/document.rs#L801)):</sup>
How's this? It still uses Root for the comparison, but the code is cleaned up a bit...

An alternative we discussed on IRC would be to add a raw pointer comparison method that doesn't require Root.
<!-- Reviewable comment -K1jDYWM-ObwyBS4SEwI:-96971074 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8232)
<!-- Sent from Reviewable.io -->

jdm commented:
Reviewed 1 of 2 files at r3.
Review status: 3 of 7 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8232)
<!-- Sent from Reviewable.io -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #8229) made this pull request unmergeable. Please resolve the merge conflicts.

glennw commented:
Reviewed 2 of 7 files at r1.
Review status: 4 of 7 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

<sup>**[components/compositing/compositor.rs, line 84 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8232#-K24Ot78-_RQ4y5l8iSz)** ([raw file](https://github.com/servo/servo/blob/68dbab0bc2e1c1e1fe6e8fc80e53cbb3a3f7b6c4/components/compositing/compositor.rs#L84)):</sup>
Can we add some comments describing what this u32 represents?
<!-- Reviewable comment -K24Ot78-_RQ4y5l8iT-:-1829287834 -->

---

<sup>**[components/compositing/windowing.rs, line 64 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8232#-K24PAyV9rImJRm2ULmb)** ([raw file](https://github.com/servo/servo/blob/68dbab0bc2e1c1e1fe6e8fc80e53cbb3a3f7b6c4/components/compositing/windowing.rs#L64)):</sup>
Should the i32 be a new type?
<!-- Reviewable comment -K24PAyV9rImJRm2ULmc:-1148877544 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8232)
<!-- Sent from Reviewable.io -->

mbrubeck commented:
Rebased and addressed review comments.
<!-- Reviewable comment -K28L_UnkUPhbO5QuUkW:1923587821 -->

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8232)
<!-- Sent from Reviewable.io -->

glennw commented:
Reviewed 2 of 7 files at r1, 1 of 2 files at r3, 4 of 7 files at r4.
Review status: 4 of 7 files reviewed at latest revision, all discussions resolved.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8232)
<!-- Sent from Reviewable.io -->

glennw commented:
@mbrubeck The compositor changes look good to me after those changes.

mbrubeck commented:
@bors-servo r=glennw

Carrying r=jdm for the script changes; aside from rebasing, the only change was that the `identifier` argument get wrapped in the TouchId newtype.

bors-servo commented:
:pushpin: Commit c7612b1 has been approved by `glennw`

<!-- @bors-servo r=glennw c7612b1c3da6b26f7c6a0411df796eb4c8c960ea -->

bors-servo commented:
:hourglass: Testing commit c7612b1 with merge 2142ec1...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/169)

jdm commented:
```
/home/servo/buildbot/slave/linux-dev/build/ports/glutin/window.rs:26:21: 26:34 error: unresolved import `script_traits::TouchEventType`. Maybe a missing `extern crate script_traits`? [E0432]
/home/servo/buildbot/slave/linux-dev/build/ports/glutin/window.rs:26 use script_traits::{TouchEventType, TouchId};
                                                                                         ^~~~~~~~~~~~~
/home/servo/buildbot/slave/linux-dev/build/ports/glutin/window.rs:26:21: 26:34 help: run `rustc --explain E0432` to see a detailed explanation
/home/servo/buildbot/slave/linux-dev/build/ports/glutin/window.rs:26:37: 27:4 error: unresolved import `script_traits::TouchId`. Maybe a missing `extern crate script_traits`? [E0432]
/home/servo/buildbot/slave/linux-dev/build/ports/glutin/window.rs:26 use script_traits::{TouchEventType, TouchId};
/home/servo/buildbot/slave/linux-dev/build/ports/glutin/window.rs:27 #[cfg(feature = "window")]
/home/servo/buildbot/slave/linux-dev/build/ports/glutin/window.rs:26:37: 27:4 help: run `rustc --explain E0432` to see a detailed explanation
error: aborting due to 2 previous errors
```

mbrubeck commented:
@bors-servo r=glennw

Moved an import to fix SERVO_HEADLESS builds.

bors-servo commented:
:pushpin: Commit ef93650 has been approved by `glennw`

<!-- @bors-servo r=glennw ef93650db944c0865da9f0c0b198dbc89fc46bd4 -->

bors-servo commented:
:hourglass: Testing commit ef93650 with merge 3fdaa6e...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/182), [gonk](http://build.servo.org/builders/gonk/builds/187), [linux-dev](http://build.servo.org/builders/linux-dev/builds/182), [linux-rel](http://build.servo.org/builders/linux-rel/builds/200), [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/195), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/193), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/204)
